crates/edr_evm/src/state/overrides.rs

```diff
@@ -247,34 +249,100 @@ impl TryFrom<StateOverrideOptions> for StateOverrides {
 }
 
 /// A wrapper around a state ref object that applies overrides.
-pub struct StateRefOverrider<'state, StateErrorT> {
-    overrides: &'state StateOverrides,
-    state: &'state dyn StateRef<Error = StateErrorT>,
+pub struct StateRefOverrider<'overrides, StateT> {
+    overrides: &'overrides StateOverrides,
+    state: StateT,
 }
 
-impl<'state, StateErrorT> StateRefOverrider<'state, StateErrorT> {
+impl<'overrides, StateT> StateRefOverrider<'overrides, StateT> {
     /// Creates a new state ref overrider.
     pub fn new(
-        overrides: &'state StateOverrides,
-        state: &'state dyn StateRef<Error = StateErrorT>,
-    ) -> StateRefOverrider<'state, StateErrorT> {
+        overrides: &'overrides StateOverrides,
+        state: StateT,
+    ) -> StateRefOverrider<'overrides, StateT> {
         StateRefOverrider { overrides, state }
     }
 }
 
-impl<'state, StateErrorT> StateRef for StateRefOverrider<'state, StateErrorT> {
-    type Error = StateErrorT;
+impl<'state, StateT: StateRef> StateRef for StateRefOverrider<'state, StateT> {
+    type Error = StateT::Error;
 
     fn basic(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
-        self.overrides.account_info(self.state, &address)
+        self.overrides.account_info(&self.state, &address)
     }
 
     fn code_by_hash(&self, code_hash: B256) -> Result<Bytecode, Self::Error> {
-        self.overrides.code_by_hash(self.state, code_hash)
+        self.overrides.code_by_hash(&self.state, code_hash)
     }
 
     fn storage(&self, address: Address, index: U256) -> Result<U256, Self::Error> {
         self.overrides
-            .account_storage_at(self.state, &address, &index)
+            .account_storage_at(&self.state, &address, &index)
     }
 }
+
+// // TODO: Remove this once trait upcasting is supported
```

Comment:
Can this and the lines below be removed?

Reply:
Addressed by https://github.com/NomicFoundation/hardhat/pull/4968/commits/12a22bf9d1642a3bf8e42dfcb18f94b522fd32d7